### PR TITLE
Graph analyser improvements

### PR DESCRIPTION
**Pull request by danielchriscarter**

### CODE REVIEW

File: macros.py [hunk 1]
```diff
@@ -164,6 +164,48 @@ def __repr__(self):
...
+        for line in lines:
+            sets[line.origin][line.dest].append(line)
+
+        result = []
+        for origin, dest_set in sets.items():
+            for dest, line_set in dest_set.items():
+                if len(line_set) == 1:
+                    result.append(line_set[0])
+                else:
+                    # Alternative layout where there are 2 vulnerabilities
```
Comment (drt24) [inline macros.py:201]: 2 or more

File: tools/graph_analyser/analyser.py [hunk 2]
```diff
@@ -42,23 +74,63 @@ def print_search(graph, start, end):
 
         # Remove duplicate edges
         sgraph = strictify(graph)
-        
+
+        if dfs(sgraph, 'remote', 'kernel'):
+            grid[vindex, dindex] = 17
```
Comment (drt24) [inline tools/graph_analyser/analyser.py:79]: What is the significance of this reverse sequence of odd numbers?

**Review by drt24 — APPROVED** [review]
Just a couple of minor points for clarification. LGTM.

File: macros.py (hunk 1 above)
Comment (danielchriscarter) [reply macros.py:201]: Apologies- badly written comment. Was meaning to say that the commented-out line below was an alternative layout to the one in use (i.e. for 2 vulnerabilities, show both names combined, and for 3 or more, just show the number). This didn't end up being very clear though, so I've just gone for showing the number any time there's more than one.

File: tools/graph_analyser/analyser.py (hunk 2 above)
Comment (danielchriscarter) [reply tools/graph_analyser/analyser.py:79]: Intended so they are unambiguously between the (even number) "limits" for each of the categories, but probably not really necessary.

Comment (drt24) [reply tools/graph_analyser/analyser.py:79]: Yes I worked that out as I read on, it is a good strategy, but a comment at this point would help clarify to the future reader.

**Pull request merged**